Add page on source code verification

emmanuel-awosika opened the pull request:
Describes the process (and importance) of verifying source code for Ethereum smart contracts.

<!--- Provide a general summary of your changes in the Title above -->

## Description

The page explains what source code verification means, how it's done, why it matters for users and developers. It also has a section describing the tools for verifying contract code for Ethereum smart contracts. 

<!--- Describe your changes in detail -->

## Related Issue

#6308

kuzdogan reviewed (commented):
This is overall great thanks for writing!

I fixed some typos and left some small comments to what you have written. The only issue is the article is lacking a little the difference between a "full match" and a "partial match", and what contract metadata is. Partial match (how Etherscan verifies currently), which is what people usually refer to with verification, does not guarantee many things such as variable names, comments, NatSpec etc. It is only a verification that this piece of code compiles to the same bytecode. There are many parts in the text that this distinction is missing and I think it is important to convey this information and push people towards full verification, as well as introduce to them why and how the contract metadata is useful.

I can suggest writing the "How" and "Sourcify" parts from scratch, adding a section for contract metadata and full matches. Would you be okay with it and would it be possible in this PR to make changes directly?

emmanuel-awosika commented on src/content/developers/docs/smart-contracts/source-code-verification.md:
```diff
@@ -0,0 +1,95 @@
+
+## What is source code verification? {#what-is-source-code-verification}
+
+A smart contract’s source code, which are instructions [written in Solidity](/developers/docs/smart-contracts/languages/) or another high-level programming language, is usually [compiled into bytecode](/developers/docs/smart-contracts/compiling/) before being deployed in the [Ethereum Virtual Machine (EVM)](/developers/docs/evm/). This happens because the EVM cannot interpret high-level instructi ...
```
Yes, thanks for pointing out. I've taken it out now.

emmanuel-awosika commented on src/content/developers/docs/smart-contracts/source-code-verification.md:
Edited the text to reflect your suggestion. Let me know your thoughts on it.

emmanuel-awosika commented on src/content/developers/docs/smart-contracts/source-code-verification.md:
```diff
@@ -0,0 +1,95 @@
+---
+
+[Smart contracts](/developers/docs/smart-contracts/) are designed to be “trustless”, meaning users shouldn’t have to trust third parties (e.g., developers and companies) before interacting with a contract. As a requisite for trustleness, users and other developers must be able to verify a smart contract’s source code. Source code verification assures users and developers that the published contra ...
+
```
Added this now.

emmanuel-awosika commented on src/content/developers/docs/smart-contracts/source-code-verification.md:
Changed to: "Contract verification can prevent this by checking if comments in the published source file match those used during compilation, although this depends on whether the contract was fully or partially verified.". Does that work?

kuzdogan replied:
I think we should better distinguish code vs. comment. Because even a malicious actor can't straightforwardly add malicious code to a verified contract, because it would change the bytecode, but comments can be freely edited with partial matches. Although there was a now fixed way to do this in Etherscan, because of how they parse the metadata hash. Check [this](https://samczsun.com/hiding-in-plain-sight/) out if you're interested.)

I'll suggest some changes to this in my PR to yours.

emmanuel-awosika commented on src/content/developers/docs/smart-contracts/source-code-verification.md:
```diff
@@ -0,0 +1,95 @@
+
+Projects reduce trust assumptions by publishing the source code of their contracts. But this leads to another problem: it is difficult to verify that the published source code matches the contract bytecode. In this scenario, the value of trustlessness is lost because users have to trust developers not to change contract code (thereby changing the bytecode) before deploying. 
+
+Source code verification tools provide guarantees that a smart contract’s source code files matches the assembly code. The result is a trustless ecosystem, where users don’t blindly trust third parties and instead verify code before depositing funds into a contract. 
```
Fixed this.

emmanuel-awosika commented on src/content/developers/docs/smart-contracts/source-code-verification.md:
```diff
@@ -0,0 +1,95 @@
+sidebar: true
+---
+
+[Smart contracts](/developers/docs/smart-contracts/) are designed to be “trustless”, meaning users shouldn’t have to trust third parties (e.g., developers and companies) before interacting with a contract. As a requisite for trustleness, users and other developers must be able to verify a smart contract’s source code. Source code verification assures users and developers that the published contra ...
```
Fixed this.

emmanuel-awosika commented:
> This is overall great thanks for writing!
> 
> I fixed some typos and left some small comments to what you have written. The only issue is the article is lacking a little the difference between a "full match" and a "partial match", and what contract metadata is. Partial match (how Etherscan verifies currently), which is what people usually refer to with verification, does not guarantee many things such as variable names, comments, NatSpec etc. It is only a verification that this piece of code compiles to the same bytecode. There are many parts in the text that this distinction is missing and I think it is important to convey this information and push people towards full verification, as well as introduce to them why and how the contract metadata is useful.
> 
> I can suggest writing the "How" and "Sourcify" parts from scratch, adding a section for contract metadata and full matches. Would you be okay with it and would it be possible in this PR to make changes directly?
> 
> Edit: I can fork from your repo and open a PR onto `emmanuel-awosika:patch-12`.

Hey @kuzdogan, thanks for the feedback. I think it's a great idea to expand on differences between full vs partial matches. I only touched on it briefly (e.g., by explaining that Etherscan doesn't verify metadata hashes), but the concept could use more explanation. 

I think you should be able to edit the PR directly. I'm still getting to grips with using GitHub, so I don't know how to merge PRs from my repo yet. That said, if you can't make changes, let me know, and I'll find a way around it.

kuzdogan commented on src/content/developers/docs/smart-contracts/source-code-verification.md:
```diff
@@ -0,0 +1,95 @@
+
+Source code verification is comparing a smart contract’s source code and the compiled bytecode used during the contract creation to detect any differences. Verifying smart contracts matters because the advertised contract code may be different from what runs on the blockchain. 
+
+Source code verification ensures that functions, values, and variables defined in the source code remain the same after the contract is compiled and deployed. Source verification also makes provision for code documentation, so that end-users know what a smart contract is designed to do. 
```
Agreed it's a good intro. The second paragraph was what I was referring to. :+1:

kuzdogan commented:
I can't edit the PR directly but a PR is basically a merge req. from one branch to another so if you make changes to the suggesting branch, the PR will update too. That's why I will be forking your repo and branch, and will open a PR onto yours. Once you merge them on `emanuel-awosika:patch-12`, it will be reflected in this PR.

emmanuel-awosika commented:
> I can't edit the PR directly but a PR is basically a merge req. from one branch to another so if you make changes to the suggesting branch, the PR will update too. That's why I will be forking your repo and branch, and will open a PR onto yours. Once you merge them on `emanuel-awosika:patch-12`, it will be reflected in this PR.

Alright, then. I'll look out for any merge requests.

emmanuel-awosika commented:
Hey @kuzdogan. I created a new folder for the source code verification page and added the image, too. Also, I replaced the external link with a link to the image in the folder. Does that work for you?

kuzdogan commented on src/content/developers/docs/smart-contracts/source-code-verification/index.md:
I think there needs to be a relative link here. Check Accounts page for the example: https://github.com/ethereum/ethereum-org-website/blob/dev/src/content/developers/docs/accounts/index.md

kuzdogan replied:
Also adding a higher resolution image: 
![source-code-verification](https://user-images.githubusercontent.com/13069972/178468962-d6cfe435-7e0c-4dd1-9242-817eee5ba2cb.png)

kuzdogan commented:
Looks good! Just one small thing about the image and we are good to go 👍 

Couple of notes for reviewers:
- The content contains a reference to the proposed "Formal Verification" page https://github.com/ethereum/ethereum-org-website/pull/6394 so would make sense to merge after that and check the URI
- I couldn't find this page in the generated Gatsby build at https://build-c0382175-68d5-49dc-9eb8-e53324852482.gtsb.io/en/developers/docs/smart-contracts/ maybe I'm missing something.

samajammin commented on src/content/developers/docs/smart-contracts/source-code-verification/index.md:
```diff
@@ -0,0 +1,99 @@
+---
```
Minor nitpick/suggestion - could we shorten this directory name to:
`src/content/developers/docs/smart-contracts/verifying/index.md`?

This would create the page `ethereum.org/en/developers/docs/smart-contracts/verifying/`, keeping it consistent with others e.g.
- https://ethereum.org/en/developers/docs/smart-contracts/testing/
- https://ethereum.org/en/developers/docs/smart-contracts/compiling/
- https://ethereum.org/en/developers/docs/smart-contracts/deploying/

kuzdogan replied:
I think it's good for consistency as you suggest. As mentioned in the content, verifying usually refers to source code verification, plus we make the distinction with the formal v. at the very beginning. So I'd say 👍

emmanuel-awosika replied:
Looks like we have a dilemma here. If we didn't have a page on formal verification, then "verifying" would be a good directory name for the page. The best suggestion I can make is using "source-verifying" for this page and "formal verifying" for the page on formal verification.

Also, @kuzdogan, I pushed a new commit to tweak the numbered list outlining instructions for verifying contracts. The initial list didn't put spaces between items, so it showed up as a paragraph in the preview.

samajammin commented:
I fixed the image reference & replaced it with the higher res version that @kuzdogan shared. I think this is now good to go! 🚀

emmanuel-awosika commented:
> I fixed the image reference & replaced it with the higher res version that @kuzdogan shared. I think this is now good to go! 🚀

Thank you, @samajammin.
